doc/design/context.md

```diff
@@ -0,0 +1,171 @@
+## Context Design
+
+`Net` is the container and controller of a set of `Operator`. Each `Operator` in `Net` has a method called `Run` to make computation. The `Run` method of `Operator` is defined as follows:
+
+```c++
+Error Operator::Run(OpContext* context);
+```
+
+And the `OpContext` is defined as follows:
+
+```c++
+Struct OpContext {
+  Scope* scope;
+  DeviceContext* device_context;
+};
+```
+
+The `Run` method will take `Variable` (containing `Tensor`) from `Scope` to make computation on certain `DeviceContext`. `DeviceContext` provides necessary runtime resources for computation, including CudaStream, cublasHandle and so on.
```

Comment:
At first, you are right, the ```Operator::Run``` will  run on the same device.

However, the ```Context``` which binds on a specific GPU card is not enough for efficient Net execution. The ```Operator::Run``` usually binds on a specific CUDA stream on a specific GPU card. 

 If we just want to implement a ```SimpleNet```, which all the operators will execute sequentially, only one CUDA stream is enough. 

But, if we also want to implement a ```DAGNet```, which the operators can execute in parallel, we must have to create several CUDA stream on a GPU card. 

So, we may have a single thread or several threads to execute an ```Net```, and each thread will hold it own CUDA stream.

There will be two ways to implement ```DAGNet```.

- We pass a  ```Context``` parameter to the ```Operator```, and the ```Context``` must have several CUDA stream.
```
struct Context {
  int device_id;
  std::vector<cudaStream_t> streams;
};

void Operator::Run(const Context& ctx);
```
And the Operator have to choose one CUDA stream to run.

- We pass ```OpContext``` to ```Operator```

```
struct OpContext {
  int device_id;
  cudaStream_t stream;
};

void Operator::Run(OpContext& ctx);
```

And the OpContext will be passed to the Operator, and may be created by many threads. And all the threads are managed and scheduled by engine.